### ISSUE TITLE
Error: Unexpected key(s) in state_dict: "logit_scale".

### ISSUE THREAD

**Issue by samin9796** [935963004/LaBraM#27]
Hi! Thank you for your great work. I am trying to load the pre-trained base model to extract feature embeddings (no fine-tuning). When I load the model, I get the following error:

`    raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
RuntimeError: Error(s) in loading state_dict for NeuralTransformerForMEM:
	Unexpected key(s) in state_dict: "logit_scale".`

Any solution? Also, do you have any pointers like which python file should I run to extract the feature embeddings from the pre-trained model?

**Comment by samin9796** [comment]
Another question: in the forward_features method, there are four dimensions getting unpacked from the input.

`batch_size, c, time_window, _ = x.size()`

What does the fourth dimension indicate?

**Comment by 935963004** [comment]
You are recommended to use run_class_finetuning.py to load the model and remove the training code of it. For the second question, the fourth dimension indicates the patch size, which is 200 in this paper.

**Comment by samin9796** [comment]
@935963004 Thank you very much!

Another question, the model is expecting the patches instead of (batch_size, num_of_channels, time_steps). I could not find the code that creates the patches in this repository. Could you kindly point to the code which does that?

**Comment by samin9796** [comment]
While using the pre-trained base model to extract features from my dataset, I get the following warning:

`size mismatch for pos_embed: copying a param with shape torch.Size([1, 129, 200]) from checkpoint, the shape in current model is torch.Size([1, 11, 200]).
size mismatch for time_embed: copying a param with shape torch.Size([1, 16, 200]) from checkpoint, the shape in current model is torch.Size([1, 25, 200]).
`
However, it produces the feature embedding. Can we use the pre-trained model for any number of patches?

**Comment by 935963004** [comment]
> @935963004 Thank you very much!
> 
> Another question, the model is expecting the patches instead of (batch_size, num_of_channels, time_steps). I could not find the code that creates the patches in this repository. Could you kindly point to the code which does that?

The 65th line of engine_for_finetuning.py

**Comment by 935963004** [comment]
> While using the pre-trained base model to extract features from my dataset, I get the following warning:
> 
> `size mismatch for pos_embed: copying a param with shape torch.Size([1, 129, 200]) from checkpoint, the shape in current model is torch.Size([1, 11, 200]). ` However, it produces the feature embedding. Can we use the pre-trained model for any number of patches/channels?

Yes. I think you'd first try to replicate finetuning on TUEV dataset using our provided code and script, and ensure everything is okay. Then you may transfer to your own data.

**Comment by neerajwagh** [comment]
> You are recommended to use run_class_finetuning.py to load the model and remove the training code of it.

Hi @935963004 Weibang -- thanks for this pointer, sharing the code, and the base model! I'm trying to extract features from the frozen base model in a new jupyter notebook (copying only the strictly necessary code from your repo and run_class_finetuning.py as you suggested). **I'm not sure if utils.load_state_dict() is picking up all the pretrained weights from the labram-base.pth checkpoint file.** 

This is the code:
```
from models.LaBraM import *
from utils.LaBraM import load_state_dict

import torch
from collections import OrderedDict

args, ds_init = get_args()
# print(args)

model = get_models(args)
# print(model)

checkpoint = torch.load(args.finetune, map_location='cpu')

print("Load ckpt from %s" % args.finetune)
checkpoint_model = None
for model_key in args.model_key.split('|'):
    if model_key in checkpoint:
        checkpoint_model = checkpoint[model_key]
        print("Load state_dict by model_key = %s" % model_key)
        break
if checkpoint_model is None:
    checkpoint_model = checkpoint
if (checkpoint_model is not None) and (args.model_filter_name != ''):
    all_keys = list(checkpoint_model.keys())
    new_dict = OrderedDict()
    for key in all_keys:
        if key.startswith('student.'):
            new_dict[key[8:]] = checkpoint_model[key]
        else:
            pass
    checkpoint_model = new_dict

state_dict = model.state_dict()
for k in ['head.weight', 'head.bias']:
    if k in checkpoint_model and checkpoint_model[k].shape != state_dict[k].shape:
        print(f"Removing key {k} from pretrained checkpoint")
        del checkpoint_model[k]

all_keys = list(checkpoint_model.keys())
for key in all_keys:
    if "relative_position_index" in key:
        checkpoint_model.pop(key)

load_state_dict(model, checkpoint_model, prefix=args.model_prefix)
```

This is the output:

```
Load ckpt from ./model_checkpoints/labram-base.pth

Load state_dict by model_key = model

Weights of NeuralTransformer not initialized from pretrained model: ['blocks.0.attn.q_bias', 'blocks.0.attn.v_bias', 'blocks.1.attn.q_bias', 'blocks.1.attn.v_bias', 'blocks.2.attn.q_bias', 'blocks.2.attn.v_bias', 'blocks.3.attn.q_bias', 'blocks.3.attn.v_bias', 'blocks.4.attn.q_bias', 'blocks.4.attn.v_bias', 'blocks.5.attn.q_bias', 'blocks.5.attn.v_bias', 'blocks.6.attn.q_bias', 'blocks.6.attn.v_bias', 'blocks.7.attn.q_bias', 'blocks.7.attn.v_bias', 'blocks.8.attn.q_bias', 'blocks.8.attn.v_bias', 'blocks.9.attn.q_bias', 'blocks.9.attn.v_bias', 'blocks.10.attn.q_bias', 'blocks.10.attn.v_bias', 'blocks.11.attn.q_bias', 'blocks.11.attn.v_bias', 'fc_norm.weight', 'fc_norm.bias']

Weights from pretrained model not used in NeuralTransformer: ['mask_token', 'pos_embed', 'lm_head.weight', 'lm_head.bias', 'norm.weight', 'norm.bias']
```

Does this look correct for feature extraction? Seems like the forward pass will use the pretrained TemporalConv weights, but block weights will be randomly initialized so final features may not be useful.
